### PR TITLE
feat: treat streaming edge cases

### PR DESCRIPTION
**Pull request by r4mmer**
### Acceptance Criteria

- During history stream queue processing, await the end of JS tasks execution at the end of each loop
  - This will make IO tasks and other tasks free to run so the queue processing does not block the main loop.
- If a sync is aborted while it waits for execution, we should abort as soon as possible to avoid executing needlessly.
- Save the capabilities set from the fullnode and check if the fullnode is history-streaming capable.

### Security Checklist
- [ ] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "level": "8.0.1",
     "lodash": "4.17.21",
     "long": "5.2.3",
+    "queue-microtask": "^1.2.3",
```
Comment (r4mmer) [inline package.json:27]: This is not a new dependency, we already have this as an indirect dependency and i'm adding as a direct deendency.

File: src/new/connection.ts [hunk 2]
```diff
@@ -60,6 +62,8 @@ class WalletConnection extends BaseConnection {
 
   streamWindowSize: number | undefined;
 
+  capabilities?: FullnodeCapability[];
```
Comment (r4mmer) [inline src/new/connection.ts:65]: When the websocket connects the fullnode sends a capabilities message which contains some feature flags of the features enabled on the fullnode instance.

Currently only the `history-streaming` flag exists.

**Review by luislhl — APPROVED** [review]

**Review by luislhl — DISMISSED** [review]

File: package-lock.json [hunk 3]
```diff
@@ -19,6 +19,7 @@
         "level": "8.0.1",
         "lodash": "4.17.21",
         "long": "5.2.3",
+        "queue-microtask": "^1.2.3",
```
Comment (tuliomir) [inline package-lock.json:22]: chore: We should use a fixed version to match the pattern we use on the other dependencies.
```suggestion
        "queue-microtask": "1.2.3",
```

File: src/new/connection.ts [hunk 4]
```diff
@@ -109,6 +114,42 @@ class WalletConnection extends BaseConnection {
...
+  }
+
+  /**
+   * If the fullnode has not sent the capabilities yet wait a while.
+   */
+  async waitCapabilities() {
+    if (this.capabilities === undefined) {
+      // Wait 2s so the fullnode has some time to send the capabilities envent
+      await new Promise<void>(resolve => {
+        setTimeout(resolve, 2000);
```
[the remark this answers is not in the document]
Comment (tuliomir) [reply src/new/connection.ts:136]: suggestion(non-blocking): If we keep this same implementation, I'd only recommend moving this `2000` to a file-wide variable or even a property in the constants/configs file, for easier configuration.

File: src/sync/stream.ts [hunk 5]
```diff
@@ -426,6 +427,12 @@ export class StreamManager extends AbortController {
       throw new Error('No abort controller on connection');
     }
 
+    // The connection with the fullnode may have been lost during the time waiting to sync
+    if (signal.aborted) {
+      this.abort();
+      throw new Error('The connection already aborted this stream');
```
Comment (tuliomir) [inline src/sync/stream.ts:433]: question(non-blocking): Does the scope of this PR allow adding a test for this abort condition?

**Review by tuliomir — APPROVED** [review]

File: src/sync/stream.ts (hunk 5 above)
Comment (r4mmer) [reply src/sync/stream.ts]: During the process to add this test I checked that this was unnecessary.
The idea was that the connection could be lost when waiting on the queue for execution, but the signal is created here.

I'll check how to create the abort controller before waiting on the queue, so if the connection is lost we can truly abort.

File: package-lock.json (hunk 3 above)
Comment (r4mmer) [reply package-lock.json]: Done

**Review by luislhl — APPROVED** [review]

**Review by tuliomir — APPROVED** [review]

**Pull request merged**